jobs/base_model_validator.py

```diff
@@ -0,0 +1,203 @@
+# coding: utf-8
+#
+# Copyright 2021 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+import datetime
+import re
+
+import apache_beam as beam
+
+from core.domain import cron_services
+from core.platform import models
+from jobs import base_model_validator_errors as errors
+import python_utils
+
+
+(base_models, user_models) = models.Registry.import_models(
+    [models.NAMES.base_model, models.NAMES.user])
+datastore_services = models.Registry.import_datastore_services()
+
+MAX_CLOCK_SKEW_SECS = datetime.timedelta(seconds=1)
+
+
+class BaseValidatorDoFn(beam.DoFn):
+    def clone(self, model, **new_values):
+        """Clones the entity, adding or overriding constructor attributes.
+
+        The cloned entity will have exactly the same property values as the
+        original entity, except where overridden. By default, it will have no
+        parent entity or key name, unless supplied.
+
+        Args:
+            model: datastore_services.Model. Model to clone.
+            **new_values: dict(str: *). Keyword arguments to override when
+                invoking the cloned entity's constructor.
+
+        Returns:
+            *. A cloned, and possibly modified, copy of self. Subclasses of
+            BaseModel will return a clone with the same type.
+        """
+        # Reference implementation: https://stackoverflow.com/a/2712401/4859885.
+        cls = model.__class__
+        props = {k: v.__get__(model, cls) for k, v in cls._properties.items()} # pylint: disable=protected-access
+        props.update(new_values)
+        return cls(id=model.id, **props)
+
+
+class ValidateModelIdWithRegex(BaseValidatorDoFn):
+    """DoFn to validate model ids against a given regex string."""
+
+    def __init__(self, regex_string):
+        """Initializes the ValidateModelIdWithRegex DoFn.
+
+        Args:
+            regex_string: str. A regex pattern for valid ids.
+        """
+        super(ValidateModelIdWithRegex, self).__init__()
+        self.regex_string = regex_string
```

Comment:
Let's compile this here, actually.

```suggestion
        self.regex = re.compile(regex_string)
```